lib/redis_memo/memoize_method.rb

```diff
@@ -81,14 +81,63 @@ def self.extract_dependencies(ref, *method_args, &depends_on)
 
   def self.get_or_extract_dependencies(ref, *method_args, &depends_on)
     if RedisMemo::Cache.local_dependency_cache
-      RedisMemo::Cache.local_dependency_cache[ref] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on][method_args] ||= extract_dependencies(ref, *method_args, &depends_on)
+      RedisMemo::Cache.local_dependency_cache[ref.class] ||= {}
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on] ||= {}
+      named_args = exclude_anonymous_args(depends_on, ref, method_args)
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on][named_args] ||= extract_dependencies(ref, *method_args, &depends_on)
     else
       extract_dependencies(ref, *method_args, &depends_on)
     end
   end
 
+  # We only look at named method parameters in the dependency block in order to define its dependent
+  # memos and ignore anonymous parameters, following the convention that nil or :_ is an anonymous parameter.
+  # Example:
+  # ```
+  #    def method(param1, param2)
+  #    end
+  #
+  #    memoize_method :method do |_, _, param2|`
+  #      depends_on RedisMemo::Memoizable.new(param2: param2)
+  #    end
+  # ```
+  #  `exclude_anonymous_args(depends_on, ref, [1, 2])` returns [2]
+  def self.exclude_anonymous_args(depends_on, ref, args)
+    return [] if depends_on.parameters.empty? or args.empty?
+
+    positional_args = []
+    kwargs = {}
+    named_rest = false
+    depends_on_args = [ref] + args
+    options = depends_on_args.extract_options!&.clone
+
+    depends_on.parameters.each_with_index do |param, i|
+      unless param.size != 2 || param.last == :_
+        # Defined by https://github.com/ruby/ruby/blob/22b8ddfd1049c3fd1e368684c4fd03bceb041b3a/proc.c#L3048-L3059
+        case param.first
+        when :opt, :req
+          positional_args << depends_on_args[i]
+        when :rest
+          named_rest = true
+          positional_args.concat(depends_on_args[i..-1])
```

Comment:
it's possible to have
```ruby
def test(a, b, *, c, d:); end
test(1,2,3,4,5,6,7,8,9,10,c:11,d:12)
```
we cannot just use the rest.

PS we should not mutate `depends_on_args` right? this looks like we have excluded the options

Reply:
I was thinking that it doesn't matter if we mutate depends_on_args, since we already created a copy of the original args when we append the reference to it `[ref] + args`.

I initially did `depends_on_args.extract_options!.clone` so that we would clone the keyword args and not mutate the original keyword args. Though alternatively, we could just create a deep copy of `depends_on_args`